sql: add infrastructure for rewriting computed column expressions

RaduBerinde opened the pull request:
sql: add infrastructure for rewriting computed column expressions
  
This commit adds infrastructure that allows automatic replacement of
computed column expressions in CREATE TABLE and ALTER TABLE
statements. This provides a workaround for cases where a computed
column expression is no longer allowed (because of better detection of
context-dependent operators) but the DDL statements cannot be easily
changed.

This is configured via a new (hidden) session setting
`experimental_computed_column_rewrites`. The default value of the
setting can be controlled using the cluster setting
`sql.defaults.experimental_computed_column_rewrites`.

The value is a string of the form:
```
(before expression) -> (after expression) [, (before expression) -> (after expression) ...]
```

In order for a computed column to be rewritten, the entire computed
column expression must parse into the same AST as a "before
expression".

Example:
```
CREATE TABLE t (ts TIMESTAMPTZ NOT NULL, shard INT NOT NULL AS (mod(fnv32(ts::STRING), 4)) STORED);
ERROR: mod(): fnv32(): timestamptz::string: context-dependent operators are not allowed in computed column
SQLSTATE: 0A000
HINT: TIMESTAMPTZ to STRING casts depend on the current timezone; consider using (t AT TIME ZONE 'UTC')::STRING instead.

SET experimental_computed_column_rewrites = "(mod(fnv32(ts::STRING), 4)) -> (mod(fnv32((ts AT TIME ZONE 'utc')::STRING), 4))";

CREATE TABLE t (ts TIMESTAMPTZ NOT NULL, shard INT NOT NULL AS (mod(fnv32(ts::STRING), 4)) STORED);

SHOW CREATE TABLE t;
  table_name |                                        create_statement
-------------+--------------------------------------------------------------------------------------------------
  t          | CREATE TABLE public.t (
             |     ts TIMESTAMPTZ NOT NULL,
             |     shard INT8 NOT NULL AS (mod(fnv32(timezone('utc':::STRING, ts)::STRING), 4:::INT8)) STORED,
             |     rowid INT8 NOT VISIBLE NOT NULL DEFAULT unique_rowid(),
             |     CONSTRAINT "primary" PRIMARY KEY (rowid ASC),
             |     FAMILY "primary" (ts, shard, rowid)
             | )
```

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/67120)
<!-- Reviewable:end -->

petermattis reviewed (commented):
One limitation of this facility is that it applies to every computed column, so you can't restrict such a rewrite to a computed column of one particular table. For the case of a `TIMESTAMPTZ::STRING` rewrite that is fine (the DDL would have been rejected), but I can imagine other use cases where you'd want to restrict where this is applied. Just noting this as a limitation for posterity. I don't think we should do anything more sophisticated in this PR. I'll leave the actual code review to others who are more familiar with this code.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67120#-:-MdXwoRR8RizCk9KBFGQ:b-cd1tvv)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @rytaft)



<!-- Sent from Reviewable.io -->

jordanlewis reviewed (commented):
I think we should consider storing the information in a structured form (JSON? arrays?) rather than string interpolation. I think this could help users understand the format better, and maybe have other benefits as well.

Also, I think it would be useful to follow the `sql.defaults` pattern, and then add a corresponding session setting for this variable, so we can conceivably avoid the issue that Peter is mentioning that it's a global hammer in all cases.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67120#-:-MdYBXZX3tZA5BxwHkNA:bttmkqv)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @rytaft)



<!-- Sent from Reviewable.io -->

RaduBerinde commented:
I will change it to a session/cluster setting pair.

> I think we should consider storing the information in a structured form (JSON? arrays?) rather than string interpolation. I think this could help users understand the format better, and maybe have other benefits as well.

Storing it where? We have to store it in a string if it's a cluster setting. Are you saying that the string should be in JSON format? Something like
```json
[
  { "before": "a+1", "after": "a+2" }
]
```
? I don't think that improves anything, and the downside is that we're now intermingling two different parsers (one for json, one for the SQL strings) - e.g. correct quoting can become (even) more confusing.

jordanlewis commented:
>  We have to store it in a string if it's a cluster setting.

The `SetClusterSettingNode` takes an `Expr`, and settings are typed so I think we could conceivably introduce a `RegisterJSONSetting` in the `pkg/settings` package that takes a JSON object from SQL.

RaduBerinde commented:
Supporting JSON cluster settings is non-trivial new infrastructure. And if you want a session/cluster setting pair, all the infrastructure around session settings involves strings so that would also be a major change. This is not something I'd be comfortable pursuing in something that we want merged and backported all the way to 20.2 ASAP.

We are all on the same page that this is not an ideal feature either way we cut it.

jordanlewis commented:
I'm okay merging without JSON, I didn't have a full picture of the urgency. If we ever make something more sophisticated we can polish it further.

RaduBerinde commented:
Switched to an `experimental_computed_column_rewrites` session setting, with `sql.defaults.experimental_computed_column_rewrites` cluster setting.

RaduBerinde committed:
sql: add infrastructure for rewriting computed column expressions

This commit adds infrastructure that allows automatic replacement of
computed column expressions in CREATE TABLE and ALTER TABLE
statements. This provides a workaround for cases where a computed
column expression is no longer allowed (because of better detection of
context-dependent operators) but the DDL statements cannot be easily
changed.

This is configured via a new (hidden) session setting
`experimental_computed_column_rewrites`. The default value of the
setting can be controlled using the cluster setting
`sql.defaults.experimental_computed_column_rewrites`.

The value is a string of the form:
```
(before expression) -> (after expression) [, (before expression) -> (after expression) ...]
```

In order for a computed column to be rewritten, the entire computed
column expression must parse into the same AST as a "before
expression".

Example:
```
CREATE TABLE t (ts TIMESTAMPTZ NOT NULL, shard INT NOT NULL AS (mod(fnv32(ts::STRING), 4)) STORED);
ERROR: mod(): fnv32(): timestamptz::string: context-dependent operators are not allowed in computed column
SQLSTATE: 0A000
HINT: TIMESTAMPTZ to STRING casts depend on the current timezone; consider using (t AT TIME ZONE 'UTC')::STRING instead.

SET experimental_computed_column_rewrites = "(mod(fnv32(ts::STRING), 4)) -> (mod(fnv32((ts AT TIME ZONE 'utc')::STRING), 4))";

CREATE TABLE t (ts TIMESTAMPTZ NOT NULL, shard INT NOT NULL AS (mod(fnv32(ts::STRING), 4)) STORED);

SHOW CREATE TABLE t;
  table_name |                                        create_statement
-------------+--------------------------------------------------------------------------------------------------
  t          | CREATE TABLE public.t (
             |     ts TIMESTAMPTZ NOT NULL,
             |     shard INT8 NOT NULL AS (mod(fnv32(timezone('utc':::STRING, ts)::STRING), 4:::INT8)) STORED,
             |     rowid INT8 NOT VISIBLE NOT NULL DEFAULT unique_rowid(),
             |     CONSTRAINT "primary" PRIMARY KEY (rowid ASC),
             |     FAMILY "primary" (ts, shard, rowid)
             | )
```

Release note: None

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice job making this undesirable addition fairly innocuous. :lemon: => :tropical_drink: 

> Reviewed 13 of 13 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67120#-:-MdZ3397167lxpK-OJ76:boxt8nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @RaduBerinde and @rytaft)

___
*[pkg/sql/catalog/schemaexpr/computed_column_rewrites.go, line 48 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/67120#-MdZ043v3T-7aNN_pi4G:-MdZ043v3T-7aNN_pi4H:b67sgtx) ([raw file](https://github.com/cockroachdb/cockroach/blob/5882b0565d16ca8d938f8ce14e7858cd334d571d/pkg/sql/catalog/schemaexpr/computed_column_rewrites.go#L48)):*
> ```Go
> 	for _, v := range set.Values {
> 		binExpr, ok := v.(*tree.BinaryExpr)
> 		if !ok || binExpr.Operator.Symbol != tree.JSONFetchVal {
> ```

Clever!
___
*[pkg/sql/logictest/testdata/logic_test/computed, line 918 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/67120#-MdZ10M79qKK1IrzVAIF:-MdZ10M85GAl70mO9oJN:bblu78r) ([raw file](https://github.com/cockroachdb/cockroach/blob/5882b0565d16ca8d938f8ce14e7858cd334d571d/pkg/sql/logictest/testdata/logic_test/computed#L918)):*
> ```
> 
> statement ok
> CREATE TABLE trewrite(k INT PRIMARY KEY, ts TIMESTAMPTZ, c STRING AS (ts::STRING) STORED, FAMILY (k,ts,c))
> ```

What is the behavior when running `CREATE TABLE trewrite_copy (LIKE trewrite INCLUDING ALL)`? What about the same statement after clearing `experimental_computed_column_rewrites`? I don't think we necessarily need to handle these cases gracefully - I think it's fine if they all fail. But it might be good to add some test cases so the behavior is defined.


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
Are cluster settings included in a backup? What happens if one of these tables is backed-up and restored on another cluster?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67120#-:-MdZ6UR-1W0rfi_kTRIS:bac5zv4)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @RaduBerinde and @rytaft)



<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
Once a table/column is created, everything is as if it was created with a `CREATE TABLE/ALTER TABLE` with the new expression. There is no trace anywhere that we did a rewrite, so everything should work as expected.

I believe backup deals with descriptors directly so that should be unaffected.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67120#-:-MdZ80ijEjfZ8XNTqIXL:b-q26lga)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner, @RaduBerinde, and @rytaft)

___
*[pkg/sql/logictest/testdata/logic_test/computed, line 918 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/67120#-MdZ10M79qKK1IrzVAIF:-MdZ7snbBPvtX9S3w8q2:b2p0grx) ([raw file](https://github.com/cockroachdb/cockroach/blob/5882b0565d16ca8d938f8ce14e7858cd334d571d/pkg/sql/logictest/testdata/logic_test/computed#L918)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

What is the behavior when running `CREATE TABLE trewrite_copy (LIKE trewrite INCLUDING ALL)`? What about the same statement after clearing `experimental_computed_column_rewrites`? I don't think we necessarily need to handle these cases gracefully - I think it's fine if they all fail. But it might be good to add some test cases so the behavior is defined.
</blockquote></details>

Added. It shouldn't matter, once the original table is created, it contains the new expression.


<!-- Sent from Reviewable.io -->

mgartner reviewed (approved):
> Reviewed 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67120)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @RaduBerinde and @rytaft)

___
*[pkg/sql/logictest/testdata/logic_test/computed, line 918 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/67120#-MdZ10M79qKK1IrzVAIF:-MdZBlNr2kM_nbyTYVHz:b-8v8233) ([raw file](https://github.com/cockroachdb/cockroach/blob/5882b0565d16ca8d938f8ce14e7858cd334d571d/pkg/sql/logictest/testdata/logic_test/computed#L918)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Added. It shouldn't matter, once the original table is created, it contains the new expression.
</blockquote></details>

Oh right. It's already been rewritten, so it won't fail. :thumbsup: 


<!-- Sent from Reviewable.io -->

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67120#-:-MdZGuy25keHiikW2zfC:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @RaduBerinde and @rytaft)



<!-- Sent from Reviewable.io -->

RaduBerinde commented:
bors r+

BramGruneir commented:
@RaduBerinde, there are no test cases involving multiple replacements.  It would probably be a good idea to add one.

rytaft commented:
@RaduBerinde is OOO this week -- can it wait until he returns?
